DEV-GUIDE.md

```diff
@@ -226,6 +303,29 @@ You can delete any build artifacts with:
 make clean
 ```
 
+### Debug an Istio container with Delve
+
+To debug an Istio container with Delve in a Kubernetes environment:
+
+* Locate the Kubernetes node on which your container is running.
```

Comment:
So this assumes you have direct access to the kubernetes node ? I don't know if all clusters allow this. I'm pretty sure kube exec will not work with scratch.

That's why I was proposing to actually base on xenial and include delve in the debug containers. 
And even maybe expose a debug port, so you could remote-debug.

Reply:
Yes, I have local clusters to work with. With gcloud, I think that you can ssh into the node as well.

I see that mixer has a Dockerfile.debug file that uses xenial (but not include delve). But I don't see pilot has debug containers.
Actually, you don't have to login to the container itself to debug. And I was able to delve attach to pilot container built with the scratch base image.

for example, the following output shows where pilot is running with the pilot debug container I built and pushed up to the docker hub and I couldn't run a "sh" since it's built with scratch
```
devuser@c2-master:~/go/src/istio.io/istio$ kubectl get pods -n istio-system -o wide
NAME                             READY     STATUS    RESTARTS   AGE       IP            NODE
istio-ca-78b58d7b5d-rmz6k        1/1       Running   1          6d        10.244.2.74   c2-minion2
istio-ingress-78545c5bc9-2b8t5   1/1       Running   1          6d        10.244.1.70   c2-minion1
istio-mixer-6f7db9c9fc-tnvdt     3/3       Running   3          6d        10.244.2.70   c2-minion2
istio-pilot-85bdbfc4dd-96q97     2/2       Running   0          2m        10.244.1.76   c2-minion1
devuser@c2-master:~/go/src/istio.io/istio$ kubectl exec -it istio-pilot-85bdbfc4dd-96q97 -n istio-system sh
Defaulting container name to discovery.
Use 'kubectl describe pod/istio-pilot-85bdbfc4dd-96q97' to see all of the containers in this pod.
rpc error: code = 13 desc = invalid header field value "oci runtime error: exec failed: container_linux.go:247: starting container process caused \"exec: \\\"sh\\\": executable file not found in $PATH\"\n"
command terminated with exit code 126
```
But I can login into the node and dlv attach to it for debugging:
```
devuser@c2-minion1:~$ ps -ef | grep pilot-discovery
root     29191 29175  2 10:45 ?        00:00:00 /usr/local/bin/pilot-discovery discovery -v 2 --admission-service istio-pilot
devuser  29452 29383  0 10:46 pts/0    00:00:00 grep --color=auto pilot-discovery
devuser@c2-minion1:~$ sudo dlv attach 29191
Type 'help' for list of commands.
(dlv) break discovery.go:477
Breakpoint 1 set at 0x1794232 for istio.io/istio/pilot/proxy/envoy.(*DiscoveryService).ListAllEndpoints() ./go/src/istio.io/istio/pilot/proxy/envoy/discovery.go:477
(dlv) c
> istio.io/istio/pilot/proxy/envoy.(*DiscoveryService).ListAllEndpoints() ./go/src/istio.io/istio/pilot/proxy/envoy/discovery.go:477 (hits goroutine(418):1 total:1) (PC: 0x1794232)
Warning: debugging optimized function
   472:         ds.ldsCache.clear()
   473: }
   474:
   475: // ListAllEndpoints responds with all Services and is not restricted to a single service-key
   476: func (ds *DiscoveryService) ListAllEndpoints(_ *restful.Request, response *restful.Response) {
=> 477:         methodName := "ListAllEndpoints"
   478:         incCalls(methodName)
   479:
   480:         services := make([]*keyAndService, 0)
   481:
   482:         svcs, err := ds.Services()
(dlv) next
> istio.io/istio/pilot/proxy/envoy.(*DiscoveryService).ListAllEndpoints() ./go/src/istio.io/istio/pilot/proxy/envoy/discovery.go:478 (PC: 0x179424d)
Warning: debugging optimized function
   473: }
   474:
   475: // ListAllEndpoints responds with all Services and is not restricted to a single service-key
   476: func (ds *DiscoveryService) ListAllEndpoints(_ *restful.Request, response *restful.Response) {
   477:         methodName := "ListAllEndpoints"
=> 478:         incCalls(methodName)
   479:
   480:         services := make([]*keyAndService, 0)
   481:
   482:         svcs, err := ds.Services()
   483:         if err != nil {
(dlv) print methodName
"ListAllEndpoints"
```

Reply:
Can you try building a scratch docker for pilot, but also add dlv binary ? And kubectl exec dlv ... ? 
I think pilot will start as PID 0 - so you may be able to attach. If that works - great. 

I know GKE allows ssh into nodes, but I think we need a generic solution, and I'm not sure if developers will have permission to ssh into nodes ( even if GKE allows admins to do that ). We need devs to be able to debug, without super-privs.

Reply:
I can give that a try as a separate effort.

Reply:
@costinm I looked at building the container with dlv, and what I found are:

- adding dlv in the container requires Go to be installed in the container as well.
- in order to install them, I can't use the base image scratch (or alpine). But ubuntu works fine to build the container
- To be able to invoke dlv the way you suggested, the container needs to run in privileged mode. Therefore, the default istio install file needs to be modified to add privileged mode
- istio source code that is used to build the debug image needs to be installed in the container as well so that source debugging is possible.
- how to include the above in the build process need to be looked at as well.

If this is something we want to pursue further, I'd prefer to create a new issue to track, and a separate PR to address it. Let me know.